Create "Adding a price identifier" (#1174)
UMAprotocol/protocol: documentation/oracle/governance/adding_a_price_identifier.md

> # Adding a Price Identifier

Reviewer: Should we add this to `map.txt`?

## What is a price identifier?

> A [price identifier](../../synthetic_tokens/glossary.md) is a natural language descriptor of the reference index determining how much collateral is needed for a token sponsor to be properly collateralized. 

Reviewer: Two comments:
1. nit: why not add the price identifier tag: `#price-identifier` to this link?
2. This definition of a price identifier seems fairly tied to the idea of a synthetic token. Should our language around DVM concepts be a little more generic?

Author: 1. Yes, I was going for that and must have gotten so distracted just trying to get the right page to show up that I forgot about the anchor. 
2. I was hesitant to introduce concepts outside of what's in the glossary (although the second paragraph tries to do a bit of that). Happy to take suggestions you or Hart might have.

Reviewer: 2. SG -- just wanted to make sure it was intentional. Up to you and @hlambur how to message DVM generality.

> A [price identifier](../../synthetic_tokens/glossary.md#price-identifier) is a natural language descriptor of the reference index determining how much collateral is needed for a token sponsor to be properly collateralized. 

Reviewer 2: @rcai1 I agree with Matt's early comments: this should be generalized and not specific to synthetic tokens. This first sentence needs to describe the concept of a price identifier as it relates to the DVM. 

Perhaps "price_identifier" shouldn't even be in the glossary in the synthetic token section--the price_identifier is really an oracle concept.

Author: Great, how does the following work? Do you have any suggested phrases? 

A price identifier... is a natural language descriptor of a reference index that the oracle will determine upon request.

Because DVM voters need to be able to vote on the value of this price identifier when disputes are raised, the DVM keeps a list of approved price identifiers. 

> An example of a price identifier is “GOLD_APR20”. 

Reviewer 2: Describe this! Remember that an outsider reading `GOLD_APR20` will prob just scratch their head: what does the price identifier here actually do? What are voters expected to vote on?

Also I would pick something spot, not with an expiry date. So go with `GOLD_USD` or something like that.

Also, I think it'll look best if the identifier is formatted in code style like `this` and not quotes like "this".

Author: All of that information should be documented in an associated UMIP. Since we are going to provide a UMIP for ETHBTC, I think we should use that as our example.

Author: I'm a bit confused as to what you're looking for. Can you give me an example? I can then translate it to reflect the ETHBTC UMIP if helpful.

Author: I see your example below, now, and have committed it.

> This price identifier would be documented in the UMIP that was approved by the UMA tokenholders. 

Reviewer 2: @mrice32 @rcai1 @allisonlu22 should we do this with our NFP identifier? Probably a good idea for us to create a UMIP with what that identifier does and actually publish it?

Author: Since we are going to provide a UMIP for ETHBTC, I think we should use that as our example.

> That UMIP would contain more information about how to calculate the price identifier. 

Reviewer: nit: prefer `determine` rather than `calculate` to emphasize the human judgement idea :)

Author: Noted! Will change.

## Who controls the list of approved price identifiers?

> For each deployment of the DVM, the list of approved price identifiers is controlled by the ```IdentifierWhitelist``` contract.

Reviewer 3: why does this have three quotes around it? wont this make it a code block? Should it not just be one tick like this:
```suggestion
For each deployment of the DVM, the list of approved price identifiers is controlled by the `IdentifierWhitelist` contract.
```

Author: In the preview, it showed up inline, not as a code block, but let me change the ticks to match.

In local and testnet deployments of the DVM, `IdentifierWhitelist` is controlled by a single private key. 
In the mainnet deployment of the DVM, `IdentifierWhitelist` is controlled by a decentralized governance process, as described below.

## Adding a price identifier to a local deployment
In a local deployment, your private key controls the `IdentifierWhitelist` contract. 

> You can therefore add any price identifier desired using the `IdentifierWhitelist.addSupportedIdentifier`, as described in step 5 of this [tutorial](../synthetic_tokens/tutorials/creating_from_truffle.md). 

Reviewer: nit: anchor to that step in the tutorial?

Author: Step 5 doesn't have a header (starting with #) - is it still possible to anchor to that section?

Reviewer: Good point. You'd have to make a heading for every step to add an anchor there.

> ## Adding a price identifier to a testnet deployment

Author: Yes, we haven't written this testnet deployment section, so I'd be happy to leave it as a comment with a TODO note.

> ## Adding a price identifier to a testnet deployment

Reviewer: SG

## Adding a price identifier to the mainnet deployment
The `IdentifierWhitelist` contract in the mainnet deployment of the UMA DVM is controlled by a decentralized governance process. 
To add a new price identifier, UMA token holders must vote and approve the identifier. 

> - Step 1: Discuss

Reviewer 2: ```suggestion
###Step 1: Discuss
```

Reviewer 2: @rcai1 should we make all the steps sub subsections?

Author: I don't think that's necessary. As an aesthetic, I also don't like the idea of it polluting the sidebar on the right with many headers. Do you agree?

If you are building with a price identifier not currently supported by the UMA DVM, you will need to propose it to the community of UMA token holders for a vote. 
You should create an UMIP in which you describe your project and the new price identifier(s) being requested. 
At this time, you do not need to provide an implementation for the addition of a new price identifier. 
For details on how to write a UMIP, please see this document. This UMIP will be discussed by members of the UMA community. 

- Step 2: Get Ready For Vote

> In order for the UMIP to move to the next stage of discussion, you should construct an off-chain transaction reflecting the implementation of the addition of a new price identifier and attach it to the UMIP. 

Author: Defer to you on that - I'm not quite clear what the difference is between attaching something to a UMIP or including it in the UMIP is.

> In order for the UMIP to move to the next stage of discussion, you should construct an off-chain transaction reflecting the implementation of the addition of a new price identifier and attach it to the UMIP. 

Reviewer: Thinking more about it, I think what you have here makes sense. In the case where it's a more complex upgrade than a price identifier addition, they may want to delay constructing the txn until the details are finalized and they're ready to move to a vote.

> In order for the UMIP to move to the next stage of discussion, you should construct an off-chain transaction reflecting the implementation of the addition of a new price identifier and attach it to the UMIP. 

Reviewer 2: ```suggestion
In order for the UMIP to move to the next stage of discussion, you should construct an off-chain transaction to add the proposed price identifier to the mainnet `IdentifierWhitelist`. This transaction should be attached to the UMIP. 
```

Author: Thanks, I have included this in the latest commit.

The UMA team will put this UMIP to a UMA tokenholder vote. 
This is achieved by the proposer address, controlled by the UMA team, proposing this UMIP to the governor contract, which initiates a vote on the UMIP. 

> The UMA team will simultaneously broadcast in other voter communities (e.g. Slack, Github) that the vote is being initiated. 

Reviewer: nit: this one is a little more nuanced, maybe `foundation and affiliates` or something?

Author: Let's go with foundation, but I think Hart should also take a look at this as well.

> The UMA foundation will simultaneously broadcast in other voter communities (e.g. Slack, Github) that the vote is being initiated. 

Reviewer 2: Now that we have a UMIP doc, I don't think this technical detail on how UMIPs get deployed should live here—this should be described in UMIPs doc. Really all step 2 should be is detail that you need to get the UMIP approved, which requires a token holder vote.

Author: Great, we can remove this detail and link to the UMIP part of the docs site instead. You'll see this in a future commit.

- Step 3: Vote

UMA voters will vote on the proposed transaction. Each UMA token represents one vote. If at least 5% of all tokens are used to vote, of which >50% of votes approve the UMIP, the UMIP is considered approved. 

- Step 4: Execute Transaction

> Once the price identifier is approved, anyone can tell the governor contract to execute the proposed transaction. 

Reviewer: ```suggestion
Once the proposal has been approved, anyone can tell the governor contract to execute the proposed transaction. 
```
